### PR TITLE
Add Anne Pro 2 keyboard

### PR DESCRIPTION
**Pull request by bwisn**
## Description

Introduces support for Anne Pro 2

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by fauxpark — CHANGES_REQUESTED** [review]
Please submit the core changes as a separate PR.

**Comment by bwisn** [comment]
> Please submit the core changes as a separate PR.
 Removed core and submitted it in another PR:
#14388

**Review by zvecr — CHANGES_REQUESTED** [review]
I would also recommend removing all non-default keymaps from the PR. Theres quite a lot that needs to change (a lot of which is on the PR checklist), which bloats this PR and would be better for them to resubmit after this has merged.

**Comment by tzarc** [comment]
Retargeted to `develop` due to dependent PR requiring core changes.

File: keyboards/annepro2/annepro2.c
Comment (drashna) [inline keyboards/annepro2/annepro2.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/annepro2/c15/config.h
Comment (drashna) [inline keyboards/annepro2/c15/config.h:57]: This should go in the keyboard's header file, not here. 
```suggestion
```

File: keyboards/annepro2/c15/config.h
Comment (drashna) [inline keyboards/annepro2/c15/config.h:149]:
```suggestion
```

File: keyboards/annepro2/c15/rules.mk
Comment (drashna) [inline keyboards/annepro2/c15/rules.mk:9]:
```suggestion
```

File: keyboards/annepro2/c15/rules.mk
Comment (drashna) [inline keyboards/annepro2/c15/rules.mk:28]: This should be converted to matrix lite, if/when possible. 

https://docs.qmk.fm/#/custom_matrix?id=lite

File: keyboards/annepro2/c18/config.h
Comment (drashna) [inline keyboards/annepro2/c18/config.h:174]:
```suggestion
```

File: keyboards/annepro2/c18/rules.mk
Comment (drashna) [inline keyboards/annepro2/c18/rules.mk:33]: Same as above (matrix lite)

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by bwisn** [comment]
@drashna fixed!

File: keyboards/annepro2/info.json
Comment (drashna) [inline keyboards/annepro2/info.json:6]:
```suggestion
```

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:4]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

And:
```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/annepro2/keymaps/default-full-caps/config.h
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

```suggestion
#pragma once
// Obins stock firmware has something similar to this already enabled, but disabled by default in QMK
```

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c
Comment (drashna) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:115]:
```suggestion

// Code to run after initializing the keyboard
void keyboard_post_init_user(void) {
    // Here are two common functions that you can use. For more LED functions, refer to the file "qmk_ap2_led.h"

    // annepro2-shine disables LEDs by default. Uncomment this function to enable them at startup.
    // annepro2LedEnable();

    // Additionally, it also chooses the first LED profile by default. Refer to the "profiles" array in main.c in
    // annepro2-shine to see the order. Replace "i" with the index of your preferred profile. (i.e the RED profile is index 0)
    // annepro2LedSetProfile(i);
}
```

File: keyboards/annepro2/keymaps/default-layer-indicators/config.h
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/config.h:2]: Same as previous

File: keyboards/annepro2/keymaps/default-layer-indicators/keymap.c
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/keymap.c:1]: Same as previous

File: keyboards/annepro2/keymaps/default-layer-indicators/keymap.c
Comment (drashna) [inline keyboards/annepro2/keymaps/default-layer-indicators/keymap.c:115]: While the name doesn't technically matter to the compiler, this is a bitmask of the layer state, so `state` is a bit more accurate and clear. 
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
    switch (get_highest_layer(state)) {
```

File: keyboards/annepro2/keymaps/default/config.h
Comment (drashna) [inline keyboards/annepro2/keymaps/default/config.h:1]: Same as previous

File: keyboards/annepro2/keymaps/default/keymap.c
Comment (drashna) [inline keyboards/annepro2/keymaps/default/keymap.c:1]: same as previous

File: keyboards/annepro2/qmk_ap2_led.c
Comment (drashna) [inline keyboards/annepro2/qmk_ap2_led.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: keyboards/annepro2/qmk_ap2_led.h
Comment (drashna) [inline keyboards/annepro2/qmk_ap2_led.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/annepro2/annepro2.h
Comment (drashna) [inline keyboards/annepro2/annepro2.h:22]: This line is not needed, at all. 
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/annepro2/keymaps/default-full-caps/keymap.c
Comment (fauxpark) [inline keyboards/annepro2/keymaps/default-full-caps/keymap.c:55]: These need to be updated.

**Comment by rpeshkov** [comment]
I have the same problem as for @tzarc. Keyboard gets flashed, but it's not functional. Previously I had success with repo https://github.com/OpenAnnePro/qmk_firmware.

Played around with the code in this PR. For me reverting back `CUSTOM_MATRIX = yes` and restoring matrix methods did the trick and my Anne Pro started to work. I have C18 version.

**Comment by tzarc** [comment]
For reference, I've left this PR alone as you've resolved comments on the GitHub UI, but not fixed them in the code.
We spoke on Discord about this, you apologised, and they're still unaddressed.

I'd suggest going through each and every comment made by myself, @drashna and @zvecr, ensuring every item has been addressed.

Feel free to ping here when done -- I'll go through every request to ensure it's been handled.

File: keyboards/annepro2/annepro2_ble.c
Comment (tzarc) [inline keyboards/annepro2/annepro2_ble.c:159]:
```suggestion
    static const uint8_t dummy[3] = {0};
    sdWrite(&SD1, dummy, sizeof(dummy));
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.c
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.c:21]: ChibiOS board definitions should not include keyboard files.
```suggestion
#include "hal.h"
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.c
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.c:119]: These are QMK-specific constructs and do not belong in the `board.c`.
You've misinterpreted the intent here -- `ht32_clock_init();` does indeed go into `__early_init()`, as per normal board files for stm32, which place `stm32_clock_init()` in the same location.

As I pointed out on specific lines highlighted in the previous comment, the issue was the `gpt_init()` and the `spi_init()`, not the placement of the clock init.

Refer to an existing ChibiOS board file for example structure.

```suggestion
void __early_init(void) { ht32_clock_init(); }

void boardInit(void) {}
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.h
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.h:64]: These are keyboard-specific, and should be swapped out for normal GPIO definitions as per the MCU's datasheet.

The config.h file for each board can then be keyboard-specific and map the rows/cols to the MCU pin definitions.

```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C15/board.h
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C15/board.h:25]:
```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C18/board.h
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C18/board.h:25]:
```suggestion
```

File: keyboards/annepro2/boards/ANNEPRO2_C18/board.h
Comment (tzarc) [inline keyboards/annepro2/boards/ANNEPRO2_C18/board.h:63]: As per C15.
```suggestion
```

File: keyboards/annepro2/c15/rules.mk
Comment (tzarc) [inline keyboards/annepro2/c15/rules.mk:4]:
```suggestion
```

File: keyboards/annepro2/c18/rules.mk
Comment (tzarc) [inline keyboards/annepro2/c18/rules.mk:4]:
```suggestion
```

File: keyboards/annepro2/chconf.h
Comment (tzarc) [inline keyboards/annepro2/chconf.h:68]: Remove as much of these as is possible.
This looks like copypaste without understanding the implications.

```suggestion
```

File: keyboards/annepro2/mcuconf.h
Comment (tzarc) [inline keyboards/annepro2/mcuconf.h:69]:
```suggestion
```

File: keyboards/annepro2/protocol.h
Comment (tzarc) [inline keyboards/annepro2/protocol.h:10]:
```suggestion
#include <stdint.h>
```

File: keyboards/annepro2/readme.md
Comment (tzarc) [inline keyboards/annepro2/readme.md:25]: This document needs flashing instructions embedded here, rather than linked to.

```suggestion
```

**Review by tzarc — CHANGES_REQUESTED** [review]
Round 2.

**Comment by bwisn** [comment]
@tzarc friendly ping :) IMO everything is fixed, waiting for review

**Comment by blaa** [comment]
We can wait, but maybe there's something more we can do. We've got a recent branch with a better RGB integration with the QMK (letting QMK control LEDs via commands to Shine/LED CPU), it still needs tests on C15 variant (less RAM) though. Our original idea was to leave it as it is, well-tested and work on bells and whistles later.

**Review by drashna — APPROVED** [review]

**Comment by bwisn** [comment]
Please change target from master to develop. It needs #16548 to be merged first.

**Review by tzarc — APPROVED** [review]
Close enough.

**Pull request merged**
